src/fsharp/Optimizer.fs

```diff
@@ -2270,6 +2270,14 @@ and OptimizeFastIntegerForLoop cenv env (spStart, v, e1, dir, e2, e3, m) =
 
             mkLdlen cenv.g (e2R.Range) arre, CSharpForLoopUp
 
+        | FSharpForLoopUp, Expr.Op (TOp.ILAsm ([ (AI_sub | AI_sub_ovf)], _), _, [Expr.Op (TOp.ILCall(_,_,_,_,_,_,_, mth, _,_,_), _, [arre], _) as lenOp; Expr.Const (Const.Int32 1, _, _)], _) 
+                  when 
+                        mth.Name = "get_Length" && (mth.DeclaringTypeRef.FullName = "System.Span`1" || mth.DeclaringTypeRef.FullName = "System.ReadOnlySpan`1") 
```

Comment:
Are these matches against plain type names strict enough?

Reply:
It looks like we do similar type-name-only checks in other areas of the compiler (https://github.com/dotnet/fsharp/blob/f6b8fa5826d4a86f6458793bfbd9858b1c31b82c/src/fsharp/TypedTreeOps.fs#L3198) already, so this is at least on part with those IMO.

Reply:
Seems to be common place in other areas of the optimizer:

https://github.com/dotnet/fsharp/blob/1be222df60206d3498643c36043770206d20add3/src/fsharp/Optimizer.fs#L1838-L1846